### ISSUE TITLE
Garbled output from multiple async `convert()` operations

### ISSUE THREAD

**Issue by timkendrick** [elad/node-imagemagick-native#118]
Thanks for making this package! I've just found one big issue, and that's when I try to convert more than around 5 images asynchronously, and then write those output buffers to the filesystem asynchronously, some of the output images seem to be unpredictably mangled. The file contents are either mashed together, or contain large swathes of empty bytes, or are just random snatches of the image, etc.

My specs:

> imagemagick-native: 1.9.0
Node: v4.3.0
ImageMagick: 6.9.3-0 Q16 x86_64 2016-02-10
OS X: v10.10.5 (14F1605)

I've tried to figure out why this happens – see below…

---

**INVESTIGATION**

I've narrowed the problem down to the buffer that is output from the async form of the `convert()` operation. If you write the output buffer to the filesystem **synchronously**, then it all goes well, but if you attempt to use it **asynchronously**, its contents will be messed up.

You can work around this by cloning the output buffer in the callback function, then performing the async write using this cloned buffer, and it's all fine.

This suggests to me that the output buffer is being modified some time after the callback fires – is the same buffer reused by subsequent `convert()` operations or something?

**EXAMPLE**

I've knocked up a simple script that demonstrates this behavior – it depends on `glob`, and expects a folder of images at `./input` and an empty directory at `./output`:

```javascript
var path = require('path');
var fs = require('fs');
var glob = require('glob');
var imagemagick = require('imagemagick-native');

var inputDir = './input';
var outputDir = './output';

var imagePaths = glob.sync('*.{jpg,png,gif}', { cwd: inputDir });

imagePaths.forEach(function(imagePath) {
	imagemagick.convert({
		srcData: fs.readFileSync(path.join(inputDir, imagePath)),
		width: 256,
		height: 256
	}, function(error, data) {
		if (error) { throw error; }
		var outputPath = path.join(outputDir, imagePath);

		// Workaround: returning a cloned buffer fixes the issue...
		var shouldCloneBuffer = false;
		if (shouldCloneBuffer) { data = new Buffer(data); }

		// ...otherwise this will produce mangled output
		fs.writeFile(outputPath, data, function(error) {
			if (error) { throw error; }
			console.info('Processed ' + imagePath);
		});

		/*
		// ALTERNATIVELY: (this also produces mangled output)
		setTimeout(function() {
			fs.writeFileSync(outputPath, data);
		}, 1000);
		*/
	});
});
```

It all works fine if you change the `shouldCloneBuffer` variable to `true`.

Points to note:
- The issue doesn't occur if you use the synchronous version of `imagemagick.convert`
- The issue still occurs if you wait for all the `convert()` operations to have completed before writing the output
- The issue even occurs if you only perform a single async operation at a time (i.e. perform each `convert()` operation one-by-one in series, and then perform each `writeFile` operation one-by-one in series)

I don't know anything about C++ or Node/native interop, so I can't tell from the source why this is happening, but let me know if I can help out with any more information.

Thanks!

**Comment by elad** [comment]
I haven't looked at the source in a while, but this smells like static buffers being used by ImageMagick. Not sure if/when I'll have time to take care of it, since I, personally, am barely using this module anymore. :/

**Comment by timkendrick** [comment]
Totally understand – I guess that's an occupational hazard of open source… Any idea if there are any active maintainers for this module?

**Comment by cartuchogl** [comment]
I checked that the problem may be related to the buffer returned is freed memory in c++.
Can you check if https://github.com/cartuchogl/node-imagemagick-native/tree/copybuffer fix your problem, sorry but I only check in node 0.12.9 on debian jessie box.

**Comment by timkendrick** [comment]
@cartuchogl – just tested that now and it works perfectly. Thanks for the quick fix!

**Comment by elad** [comment]
@timkendrick yes, I have made it a policy to give contributor access to anyone whose pull requests gets merged. Obviously, @cartuchogl is one of those fine folks. :)

**Comment by elad** [comment]
Fixed by #119

**Comment by timkendrick** [comment]
That's amazing – thanks for the fix @cartuchogl and great contributor policy @elad. Works like a charm.

Cheers for the super-fast turnaround guys!
